evalml/pipelines/component_graph.py

```diff
@@ -0,0 +1,310 @@
+            component_class = handle_component_class(component)
+            component_name = component_class.name
+
+            component_dict[component_name] = [component_class]
+            if idx != 0:
+                component_dict[component_name].append(f"{handle_component_class(component_list[idx-1]).name}.x")
+        return cls(component_dict, random_state=random_state)
+
+    def instantiate(self, parameters):
+        """Instantiates all uninstantiated components within the graph using the given parameters. An error will be
+        raised if a component is already instantiated but the parameters dict contains arguments for that component.
+
+        Arguments:
+            parameters (dict): Dictionary with component names as keys and dictionary of that component's parameters as values.
+                               An empty dictionary {} implies using all default values for component parameters.
+        """
+        for component_name, component_info in self.component_dict.items():
+            component_parameters = parameters.get(component_name, {})
+            component_class = component_info[0]
+            if isinstance(component_class, ComponentBase):
+                if component_parameters != {}:
+                    raise ValueError(f"Attempting to instantiate component {component_name} with parameters {component_parameters} when component already instantiated")
+                else:
+                    continue
+
+            component_class = handle_component_class(component_class)
+
+            try:
+                new_component = component_class(**component_parameters, random_state=self.random_state)
+            except (ValueError, TypeError) as e:
+                err = "Error received when instantiating component {} with the following arguments {}".format(component_name, component_parameters)
+                raise ValueError(err) from e
+
+            self.component_dict[component_name][0] = new_component
+        return self
+
+    def compute_final_features(self, X, y=None, fit=False):
+        """Transforms the data by applying all components.
+
+        Arguments:
+            X (pd.DataFrame): Input data to the pipeline to transform.
+            y (pd.Series): The target training data of length [n_samples]
+            fit (bool): Whether to fit the estimators as well as transform it.
+                        Defaults to False.
+
+        Returns:
+            pd.DataFrame and/or pd.Series - Output(s) of the final component
+        """
+        output_cache = {}
+        final_component = None
+        for component_name in list(self._compute_order):
+            final_component = component_name
+            component_class = self.component_dict[component_name][0]
+            x_inputs = []
+            y_input = None
+            for parent_input in self.get_parents(component_name):
+                if parent_input[-2:] == '.y':
+                    y_input = output_cache[parent_input]
```

Comment:
Discussed with @eccabay : let's add check that we get error if there are multiple `y` parents here.